File: sdk/typescript/src/Image.ts
/**
 * Image is a fluent Dockerfile builder. The shape mirrors the Daytona SDK's
 * Image API so examples port across by changing the import. The output is a
 * Dockerfile string accessible via `.dockerfile` — the SDK ships it to
 * `POST /v2/images/build` when an Image is passed to `MicroVM.create`.
 *
 * Caller-side context (COPY/ADD from local files) needs the
 * SB_IMAGE_BUILD_CONTEXT_ENABLED operator flag on the daemon. Without that,
 * use only `runCommands`, `env`, `workdir`, `entrypoint`, `cmd`, `user`, and
 * `expose` — every step they emit is RUN/ENV/WORKDIR/etc. against the base
 * image, no local files involved.
 */
export class Image {
  private _dockerfile = "";

  private constructor(dockerfile: string) {
    this._dockerfile = dockerfile;
  }

  get dockerfile(): string {
    return this._dockerfile;
  }

  /** Start a new image from `FROM <image>`. */
  static base(image: string): Image {
    if (typeof image !== "string" || image.trim() === "") {
      throw new TypeError("Image.base requires a non-empty image string");
    }
    return new Image(`FROM ${image.trim()}\n`);
  }

  /**
   * Build from a raw Dockerfile string. Use when you need full control
   * (multi-stage builds, ARG, ONBUILD, HEALTHCHECK, etc.) that the fluent
   * API doesn't expose. Caller is responsible for the Dockerfile validity.
   */
  static fromDockerfile(dockerfile: string): Image {
    if (typeof dockerfile !== "string" || dockerfile.trim() === "") {
      throw new TypeError("Image.fromDockerfile requires a non-empty Dockerfile string");
    }
    return new Image(dockerfile.endsWith("\n") ? dockerfile : dockerfile + "\n");
  }

  /**
   * Append one or more RUN steps. Each argument becomes its own RUN line;
   * an array argument is joined with `&&` so the commands share one layer.
   */
  runCommands(...commands: (string | string[])[]): Image {
    for (const entry of commands) {
      if (Array.isArray(entry)) {
        const joined = entry.map((c) => c.trim()).filter(Boolean).join(" && ");
        if (joined !== "") {
          this._dockerfile += `RUN ${joined}\n`;
        }
        continue;
      }
      const cmd = entry.trim();
      if (cmd !== "") {
        this._dockerfile += `RUN ${cmd}\n`;
      }
    }
    return this;
  }

  /** Append ENV K=V pairs (one ENV line per call, all vars on it). */
  env(envVars: Record<string, string>): Image {
    const parts = Object.entries(envVars).map(([k, v]) => `${k}=${dockerQuote(v)}`);
    if (parts.length > 0) {
      this._dockerfile += `ENV ${parts.join(" ")}\n`;
    }
    return this;
  }

  workdir(dirPath: string): Image {
    if (dirPath.trim() === "") {
      throw new TypeError("Image.workdir requires a non-empty path");
    }
    this._dockerfile += `WORKDIR ${dirPath}\n`;
    return this;
  }

  entrypoint(entrypointCommands: string[]): Image {
    this._dockerfile += `ENTRYPOINT ${jsonExecForm(entrypointCommands)}\n`;
    return this;
  }

  cmd(cmd: string[]): Image {
    this._dockerfile += `CMD ${jsonExecForm(cmd)}\n`;
    return this;
  }

  user(username: string): Image {
    if (username.trim() === "") {
      throw new TypeError("Image.user requires a non-empty username");
    }
    this._dockerfile += `USER ${username}\n`;
    return this;
  }

  expose(port: number): Image {
    if (!Number.isInteger(port) || port < 1 || port > 65535) {
      throw new RangeError(`Image.expose: port ${port} is out of range`);
    }
    this._dockerfile += `EXPOSE ${port}\n`;
    return this;
  }
}

function dockerQuote(value: string): string {
  // Quote when the value contains whitespace or a special char; otherwise
  // leave bare to match canonical Dockerfile style.
  if (/^[A-Za-z0-9_\-./:@]+$/.test(value)) {
    return value;
  }
  return `"${value.replace(/\\/g, "\\\\").replace(/"/g, '\\"')}"`;
}

function jsonExecForm(parts: string[]): string {
  return JSON.stringify(parts);
}

File: sdk/rust/src/image.rs
use std::fmt::Write as _;

#[derive(Debug, Clone)]
pub struct Image {
    dockerfile: String,
    error: Option<String>,
}

impl Image {
    pub fn base(image: &str) -> Self {
        let trimmed = image.trim();
        if trimmed.is_empty() {
            return Self::with_error("Image::base requires a non-empty image string");
        }
        Self {
            dockerfile: format!("FROM {}\n", trimmed),
            error: None,
        }
    }

    pub fn from_dockerfile(dockerfile: &str) -> Self {
        if dockerfile.trim().is_empty() {
            return Self::with_error(
                "Image::from_dockerfile requires a non-empty Dockerfile string",
            );
        }
        let mut value = dockerfile.to_string();
        if !value.ends_with('\n') {
            value.push('\n');
        }
        Self {
            dockerfile: value,
            error: None,
        }
    }

    pub fn dockerfile(&self) -> &str {
        &self.dockerfile
    }

    pub(crate) fn validation_error(&self) -> Option<&str> {
        self.error.as_deref()
    }

    pub fn run_command(mut self, command: &str) -> Self {
        if self.error.is_some() {
            return self;
        }
        let trimmed = command.trim();
        if !trimmed.is_empty() {
            let _ = writeln!(self.dockerfile, "RUN {}", trimmed);
        }
        self
    }

    pub fn run_commands<I, S>(mut self, commands: I) -> Self
    where
        I: IntoIterator<Item = S>,
        S: AsRef<str>,
    {
        if self.error.is_some() {
            return self;
        }
        for command in commands {
            let trimmed = command.as_ref().trim();
            if !trimmed.is_empty() {
                let _ = writeln!(self.dockerfile, "RUN {}", trimmed);
            }
        }
        self
    }

    pub fn run_command_group<I, S>(mut self, commands: I) -> Self
    where
        I: IntoIterator<Item = S>,
        S: AsRef<str>,
    {
        if self.error.is_some() {
            return self;
        }
        let joined = commands
            .into_iter()
            .map(|command| command.as_ref().trim().to_string())
            .filter(|command| !command.is_empty())
            .collect::<Vec<_>>()
            .join(" && ");
        if !joined.is_empty() {
            let _ = writeln!(self.dockerfile, "RUN {}", joined);
        }
        self
    }

    pub fn env<I, K, V>(mut self, env_vars: I) -> Self
    where
        I: IntoIterator<Item = (K, V)>,
        K: AsRef<str>,
        V: AsRef<str>,
    {
        if self.error.is_some() {
            return self;
        }
        let mut pairs = env_vars
            .into_iter()
            .map(|(key, value)| (key.as_ref().to_string(), value.as_ref().to_string()))
            .collect::<Vec<_>>();
        pairs.sort_by(|left, right| left.0.cmp(&right.0));
        if !pairs.is_empty() {
            let joined = pairs
                .into_iter()
                .map(|(key, value)| format!("{}={}", key, docker_quote(&value)))
                .collect::<Vec<_>>()
                .join(" ");
            let _ = writeln!(self.dockerfile, "ENV {}", joined);
        }
        self
    }

    pub fn workdir(mut self, dir_path: &str) -> Self {
        if self.error.is_some() {
            return self;
        }
        if dir_path.trim().is_empty() {
            return self.record_error("Image::workdir requires a non-empty path");
        }
        let _ = writeln!(self.dockerfile, "WORKDIR {}", dir_path);
        self
    }

    pub fn entrypoint<I, S>(mut self, commands: I) -> Self
    where
        I: IntoIterator<Item = S>,
        S: AsRef<str>,
    {
        if self.error.is_some() {
            return self;
        }
        match json_exec_form(commands) {
            Ok(payload) => {
                let _ = writeln!(self.dockerfile, "ENTRYPOINT {}", payload);
                self
            }
            Err(err) => self.record_error(format!("encode entrypoint: {}", err)),
        }
    }

    pub fn cmd<I, S>(mut self, commands: I) -> Self
    where
        I: IntoIterator<Item = S>,
        S: AsRef<str>,
    {
        if self.error.is_some() {
            return self;
        }
        match json_exec_form(commands) {
            Ok(payload) => {
                let _ = writeln!(self.dockerfile, "CMD {}", payload);
                self
            }
            Err(err) => self.record_error(format!("encode cmd: {}", err)),
        }
    }

    pub fn user(mut self, username: &str) -> Self {
        if self.error.is_some() {
            return self;
        }
        if username.trim().is_empty() {
            return self.record_error("Image::user requires a non-empty username");
        }
        let _ = writeln!(self.dockerfile, "USER {}", username);
        self
    }

    pub fn expose(mut self, port: u32) -> Self {
        if self.error.is_some() {
            return self;
        }
        if !(1..=65535).contains(&port) {
            return self.record_error(format!("Image::expose: port {} is out of range", port));
        }
        let _ = writeln!(self.dockerfile, "EXPOSE {}", port);
        self
    }

    fn with_error(message: &str) -> Self {
        Self {
            dockerfile: String::new(),
            error: Some(message.to_string()),
        }
    }

    fn record_error(mut self, message: impl Into<String>) -> Self {
        if self.error.is_none() {
            self.error = Some(message.into());
        }
        self
    }
}

fn docker_quote(value: &str) -> String {
    if value
        .chars()
        .all(|ch| ch.is_ascii_alphanumeric() || matches!(ch, '_' | '-' | '.' | '/' | ':' | '@'))
    {
        return value.to_string();
    }
    format!("\"{}\"", value.replace('\\', "\\\\").replace('"', "\\\""))
}

fn json_exec_form<I, S>(parts: I) -> Result<String, serde_json::Error>
where
    I: IntoIterator<Item = S>,
    S: AsRef<str>,
{
    let values = parts
        .into_iter()
        .map(|part| part.as_ref().to_string())
        .collect::<Vec<_>>();
    serde_json::to_string(&values)
}
